quinn-udp/src/unix.rs

```diff
@@ -68,11 +67,12 @@ impl UdpSocket {
     ) -> Poll<Result<usize, io::Error>> {
         loop {
             let last_send_error = &mut self.last_send_error;
-            let mut guard = ready!(self.io.poll_write_ready(cx))?;
-            if let Ok(res) =
-                guard.try_io(|io| send(state, io.get_ref(), last_send_error, transmits))
-            {
-                return Poll::Ready(res);
+            ready!(self.io.poll_write_ready(cx))?;
+            match send(state, self.io.get_ref(), last_send_error, transmits) {
+                Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    self.io.clear_write_ready(cx);
```

Comment:
What if the socket becomes ready in between the call to `poll_write_ready` and `clear_write_ready`? Could this miss a wakeup? Similarly for reads.

Reply:
First, the clear_write_ready is a no-op on all implementations but tokio+unix (using AsyncFd), so we are only talking about this platform now. async-io doesn't require (or have a way of) clearing readiness events. tokio UdpSocket automatically clears readiness events for us in the call to `try_recv_from`, and it also has no way to manually do that.

The code here is following the instructions from the tokio documentation:

> Indicates to tokio that the file descriptor is no longer ready. The internal readiness flag will be cleared, and tokio will wait for the next edge-triggered readiness notification from the OS.
> It is critical that this function not be called unless your code actually observes that the file descriptor is not ready. Do not call it simply because, for example, a read succeeded; it should be called when a read is observed to block.

When we observe that the file descriptor is not ready (I/O operation returns WouldBlock), we clear the readiness status.
I don't know what would the alternative? This is exactly how tokio documentation describes it and how tokio does it itself internally.

Reply:
Thanks; I'll have to study this in some detail...

Reply:
Your concern is indeed very justified:
https://github.com/microsoft/msquic/issues/2654

Reply:
Discussed with upstream tokio; this will miss wakeups. In particular, the implementation of `clear_write_ready` in terms of `poll_write_ready` will cause readiness events that occur after the initial `poll_write_ready` to be lost.

Our basic problem is that `AsyncUdpSocket` is trying to be too low-level. Rather than trying to make the methods of `quinn_udp::UdpSocket` abstract over the underlying async primitive, let's try converting the (already object-safe) `quinn_udp::UdpSocket` itself into a trait object, with separate implementations for tokio and async-std. Those separate implementations should then call into shared code to do the OS-specific heavy-lifting. We can then hoist the trait machinery up into the `quinn` crate and leave the shared OS-specific heavy lifting code in quinn_udp, concentrating awareness of async runtimes in one place.

Reply:
I've implemented the approach I outlined above in https://github.com/quinn-rs/quinn/pull/1364.